
---
title: "Beta cells: DEG rs3842753"
date: "2024-5-15" 
date-modified: last-modified
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    warning = FALSE,
    message = FALSE
)
```


```{r }
#| code-fold: true
#| code-summary: "Setup"
suppressPackageStartupMessages({
    library(tidyseurat)
    library(Seurat)
    library(targets)
    library(ggplot2)
    library(ggrepel)
    library(ggdist)
    library(ggsignif)
    library(ggpubr)
    require(showtext)
    font_add_google(name = "Roboto", family = "Roboto")
    font_add_google(name = "Roboto Condensed", family = "Roboto Condensed")
    showtext_auto()
})

tar_config_set(store = "/scratch/domeally/DCD.tienhoven_scRNAseq.2024/_targets")
tar_source("/scratch/domeally/DCD.tienhoven_scRNAseq.2024/R")
cpus <- hprcc::slurm_allocation()$CPUs

tar_load(seurat_object_sct_annotated)
seurat_object <- load_seurat(seurat_object_sct_annotated)

seurat_object <- seurat_object |>
    dplyr::filter(diabetes_status == "NODM")

make_plots_clickable()
```

## Differential expression

```{r }
bulk <- AggregateExpression(
    seurat_object,
    return.seurat = TRUE,
    assays = "RNA",
    group.by = c("cell_type", "orig.ident", "rs3842753_consensus")
) |>
    separate(.cell, into = c("cell_type", "orig.ident", "rs3842753_consensus"), sep = "_", remove = FALSE)

beta.bulk <- subset(bulk, cell_type == "Beta")
```

::: {.panel-tabset}

## DESeq2

```{r }
Idents(beta.bulk) <- "rs3842753_consensus"

de_markers <- FindMarkers(beta.bulk,
    ident.2 = "TG", ident.1 = "TT", test.use = "DESeq2", verbose = FALSE, min.pct = 0.25, pseudocount.use = 0.001
)

de_markers$gene <- rownames(de_markers)

volcano_plot(
    .data = de_markers,
    .transcripts = "gene",
    pvalue = "p_val",
    FDR = "p_val_adj",
    FDR_cutoff = 0.05,
    log2FoldChange = "avg_log2FC",
    log2FoldChange_cutoff = 1,
    title = "Beta cells: rs3842753",
    subtitle = "Upregulated in (<- TG) susceptible vs protected (TT ->) cells",
)
```

### GSEA

```{r }
gsea_results <- perform_gsea(
    .data = de_markers,
    title = "GSEA Results for rs3842753 protected (TT) v susceptible (TG) beta cells",
    .transcripts = "gene",
    .log2fc = "avg_log2FC",
    log2fc_threshold = 0,
    analysis_type = "GO",
    ont = "BP"
)
gsea_results$plot
```

```{r }
gsea_results <- perform_gsea(
    .data = de_markers,
    title = "GSEA Results for rs3842753 protected (TT) v susceptible (TG) beta cells",
    .transcripts = "gene",
    .log2fc = "avg_log2FC",
    log2fc_threshold = 0,
    analysis_type = "KEGG"
)
gsea_results$plot
```

## Wilcoxon rank sum test

```{r }
de_markers <- FindMarkers(beta.bulk,
    ident.2 = "TG", ident.1 = "TT", test.use = "wilcox", verbose = FALSE, min.pct = 0.25, pseudocount.use = 0.001
)

de_markers$gene <- rownames(de_markers)

volcano_plot(
    .data = de_markers,
    .transcripts = "gene",
    pvalue = "p_val",
    FDR = "p_val_adj",
    FDR_cutoff = 0.05,
    log2FoldChange = "avg_log2FC",
    log2FoldChange_cutoff = 1,
    title = "Beta cells: rs3842753",
    subtitle = "Upregulated in (<- TG) susceptible vs protected (TT ->) cells",
)
```

### GSEA

```{r }
gsea_results <- perform_gsea(
    .data = de_markers,
    title = "GSEA Results for rs3842753",
    .transcripts = "gene",
    .log2fc = "avg_log2FC",
    log2fc_threshold = 0,
    analysis_type = "GO",
    ont = "BP"
)
gsea_results$plot
```

```{r }
gsea_results <- perform_gsea(
    .data = de_markers,
    title = "GSEA Results for rs3842753",
    .transcripts = "gene",
    .log2fc = "avg_log2FC",
    log2fc_threshold = 0,
    analysis_type = "KEGG"
)
gsea_results$plot
```

::::
