
---
title: "Beta cells: rs689"
date: "2024-6-1" 
date-modified: last-modified
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    warning = FALSE,
    message = FALSE
)
```


```{r }
#| code-fold: true
#| code-summary: "Setup"
suppressPackageStartupMessages({
    library(tidyseurat)
    library(Seurat)
    library(targets)
    library(ggplot2)
    library(ggdist)
    library(ggsignif)
    library(ggpubr)
})

tar_config_set(store = "/scratch/domeally/DCD.tienhoven_scRNAseq.2024/_targets")
tar_source("/scratch/domeally/DCD.tienhoven_scRNAseq.2024/R")
cpus <- hprcc::slurm_allocation()$CPUs

tar_load(seurat_object_lognorm_annotated)
seurat_object <- load_seurat(seurat_object_lognorm_annotated)

cell_types <- seurat_object |>
    dplyr::select(cell_type) |>
    dplyr::distinct()

cell_type <- "Beta"

seurat_object <- seurat_object |>
    dplyr::filter(diabetes_status == "NODM") |>
    dplyr::filter(cell_type == cell_type) 
```

::: {.panel-tabset}

## Per cell

```{r }
#| fig-width: 10
#| fig-height: 10
#| fig-cap: "Boxplots of ER gene expression, indicating median values per cell. Non-expressing cells are excluded."

er_expression <- seurat_object %>%
    dplyr::mutate(rs689_consensus = factor(rs689_consensus)) %>%
    FetchData(vars = c("rs689_consensus", er_genes_of_interest), layer = "data") |>
    pivot_longer(
        cols = -rs689_consensus,
        names_to = "gene",
        values_to = "expression"
    ) |>
    dplyr::filter(expression > 0)

ggplot(er_expression, aes(x = rs689_consensus, y = expression, fill = rs689_consensus, col = rs689_consensus)) +
    stat_dots(position = "dodgejust") +
    stat_pointinterval(col = "black", alpha = 0.6) +
    facet_wrap(~gene, scales = "free_y") +
    geom_signif(
        test = "wilcox.test",
        comparisons = list(c("AA", "TT")),
        map_signif_level = TRUE,
        margin_top = -0.4,
        tip_length = 0,
        extend_line = -0.35,
        color = "black"
    ) +
    stat_summary(
        fun = median,
        geom = "label",
        aes(label = sprintf("%.2f", after_stat(y)), alpha = 0.4),
        color = "white",
        hjust = 1.1,
        vjust = 1.1
    ) +
    theme_classic() +
    NoLegend() +
    labs(
        title = "Expression of ER Genes by rs689, per cell",
        x = "Genotype",
        y = "Normalized Expression"
    ) +
    scale_fill_manual(values = c("AA" = "darkgoldenrod2", "TT" = "darkslateblue")) +
    scale_color_manual(values = c("AA" = "darkgoldenrod2", "TT" = "darkslateblue"))
```

## Per donor

```{r }
#| fig-width: 10
#| fig-height: 10
#| fig-cap: "Boxplots of ER gene expression, indicating median aggregate expression per donor"

er_expression_PsB <- seurat_object %>%
    dplyr::mutate(rs689_consensus = factor(rs689_consensus)) %>%
    aggregate_cells("orig.ident", assays = "RNA", layers = "data") |>
    dplyr::filter(.feature %in% er_genes_of_interest) |>
    select(gene = ".feature", expression = "RNA", rs689_consensus)

ggplot(er_expression_PsB, aes(x = rs689_consensus, y = expression, fill = rs689_consensus, col = rs689_consensus)) +
    stat_dots(position = "dodgejust") +
    stat_pointinterval(col = "black") +
    facet_wrap(~gene, scales = "free_y") + 
    geom_signif(
        test = "wilcox.test",
        comparisons = list(c("TT", "TG")), 
        map_signif_level = TRUE,
        margin_top = -0.4,
        tip_length = 0,
        extend_line = -0.35,
        color = "black"
    ) +
    stat_summary(
        fun = median,
        geom = "label",
        aes(label = sprintf("%.2f", after_stat(y)), alpha = 0.4),
        color = "white",
        hjust = -0.2,
        vjust = -0.7
    ) +
    theme_classic() +
    NoLegend() +
    labs(
        title = "Expression of ER Genes by rs689, per donor",
        x = "Genotype",
        y = "Normalized Expression"
    ) +
    scale_fill_manual(values = c("AA" = "darkgoldenrod2", "TT" = "darkslateblue")) +
    scale_color_manual(values = c("AA" = "darkgoldenrod2", "TT" = "darkslateblue"))
```

::::

AA = Protected
TT = Susceptible