
---
title: "Cell wish list"
date: "2024-3-25" 
date-modified: last-modified
format:
  html:
    embed-resources: true
    code-fold: true
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    warning = FALSE,
    message = FALSE
)
```

```{r }
#| code-fold: true
#| code-summary: "Show the code"
#| cache: false
suppressPackageStartupMessages({
    library(tidyverse)
    library(targets)
    library(ggplot2)
    library(Seurat)
    library(patchwork)
    library(scCustomize)
    library(ggpubr)
    library(ComplexUpset)
    library(ggrepel)
    library(gt)
    library(Nebulosa)
})

tar_config_set(store = "/scratch/domeally/DCD.tienhoven_scRNAseq.2024/_targets")
tar_source("/scratch/domeally/DCD.tienhoven_scRNAseq.2024/R")

tar_load(c(pancdb_metadata_gt, merged_seurat_sketch_750, integrated_seurat_sketch_750, azimuth_mapped_seurat_objects, cell_cycle_csv, tosti_cell_type_csv, cluster_merged_sketch_csv))

# all_samples_seurat_object <- merged_seurat_sketch_750 |> load_seurat()
# # all_samples_seurat_object <- integrated_seurat_sketch_750 |> load_seurat()
# # # Sample metadata -------------------------------------------------------------------------------
# keep_cols <- rlang::exprs(sample_name, sample_sex, sample_age, sample_ethnicity, tissue_source, ab_positive, diabetes_status, batch, protected, dplyr::contains("consensus"))
# meta_data <- pancdb_metadata_gt |>
#     dplyr::select(!!!keep_cols)

# all_samples_seurat_object <- scCustomize::Add_Sample_Meta(seurat_object = all_samples_seurat_object, meta_data = meta_data, join_by_seurat = "orig.ident", join_by_meta = "sample_name")
# # ----------------------------------------------------------------------------------------------------
# # Cell annotations
# azimuth_data <- sub("\\.qs", "\\.csv", azimuth_mapped_seurat_objects) |> readr::read_csv(show_col_types = FALSE, progress = FALSE)
# cell_cycle_data <- cell_cycle_csv |> readr::read_csv(show_col_types = FALSE, progress = FALSE)
# gpt_cell_type_data <- "/home/domeally/workspaces/DCD.tienhoven_scRNAseq.2024/analysis_cache/gpt_cell_type_out/merged_sketch_750.csv" |> readr::read_csv(show_col_types = FALSE, progress = FALSE) |> mutate(gpt_cell_type = tolower(gpt_cell_type), gpt_cell_type = ifelse(gpt_cell_type == "pp", "gamma", gpt_cell_type))
# tosti_cell_type_data <- tosti_cell_type_csv |> readr::read_csv(show_col_types = FALSE, progress = FALSE)
# # Clusters shared with gpt-cell-type
# # clusters_sketch_data <- cluster_merged_sketch_csv |>
# #     readr::read_csv(show_col_types = FALSE, progress = FALSE)

# cell_metadata <- azimuth_data |>
#     left_join(cell_cycle_data, by = "cell") |>
#     left_join(gpt_cell_type_data, by = "cell") |>
#     left_join(tosti_cell_type_data, by = "cell") |>
#     # left_join(clusters_sketch_data, by = "cell") |>
#     select(c(cell, gpt_cell_type_sketch = "gpt_cell_type", azimuth_cell_type = "predicted.annotation.l1", tosti_cell_type = "labels", cell_cycle_phase = "Phase", seurat_clusters)) |>
#     column_to_rownames(var = "cell")
# all_samples_seurat_object <- Seurat::AddMetaData(all_samples_seurat_object, cell_metadata)
# # # ----------------------------------------------------------------------------------------------------

# #--- NODM ----
# Idents(all_samples_seurat_object) <- "diabetes_status"
# seurat_object_NODM <- subset(all_samples_seurat_object, idents = "NODM") |> JoinLayers()
# seurat_object_NODM <- seurat_object_NODM |>
#     NormalizeData(normalization.method = "LogNormalize", scale.factor = 10000) %>%
#     FindVariableFeatures(selection.method = "vst", nfeatures = 2000) %>%
#     ScaleData() %>%
#     RunPCA(features = VariableFeatures(object = .)) %>%
#     RunUMAP(dims = 1:20)
# # -------- SAVE OBJECTS --------------------------------
# qs::qsave(all_samples_seurat_object, glue::glue("{analysis_cache}/data/all_samples_seurat_object_for_cellwishlist_tmp.qs"))
# qs::qsave(seurat_object_NODM, glue::glue("{analysis_cache}/data/NODM_seurat_object_for_cellwishlist_tmp.qs"))
seurat_object <- glue::glue("{analysis_cache}/data/NODM_seurat_object_for_cellwishlist_tmp.qs") |> load_seurat()
# seurat_object_all <- glue::glue("{analysis_cache}/data/all_samples_seurat_object_for_cellwishlist_tmp.qs") |> load_seurat()
```

## Cell types

::: {.panel-tabset}

### Tosti

```{r}
#| fig-cap: "Tosti et al cell types"
#| fig-height: 10
#| fig-width: 5
scCustomize::Stacked_VlnPlot(seurat_object, features = cell_type_markers, x_lab_rotate = TRUE, colors_use = cell_type_palette, group.by = "tosti_cell_type") & labs(title = "Tosti et al 2021")
```

Cell types after [Tosti et al. (2020)](https://www.gastrojournal.org/article/S0016-5085(20)35399-3/fulltext)

### Azimuth

```{r}
#| fig-cap: "Azimuth cell types"
#| fig-height: 10
#| fig-width: 5
scCustomize::Stacked_VlnPlot(seurat_object, features = cell_type_markers, x_lab_rotate = TRUE, colors_use = cell_type_palette, group.by = "azimuth_cell_type") & labs(title = "Azimuth")
```

Cell types from [Azimuth, Human Pancreas Reference Explorer](https://azimuth.hubmapconsortium.org/references/#Human%20-%20Pancreas)

### GPT cell type

```{r}
#| fig-cap: "GPT cell types"
#| fig-height: 10
#| fig-width: 5
scCustomize::Stacked_VlnPlot(seurat_object, features = cell_type_markers, x_lab_rotate = TRUE, colors_use = cell_type_palette, group.by = "gpt_cell_type_sketch") & labs(title = "GPT")
```

Cell types from [GPTCellType](https://www.nature.com/articles/s41592-024-02235-4)

::: 

ChatGPT could be improved; could also implement marker genes + gating per Wang et al

## Donor genotypes by diabetes status

```{r}
#| eval: true
meta_data <- seurat_object[[]] |>
    select(c("orig.ident", "sample_sex", "sample_age", "sample_ethnicity", "ab_positive", "diabetes_status", "protected", "batch", "rs3842752_consensus", "rs3842753_consensus", "rs689_consensus")) |>
    unique()

meta_data_long <- meta_data %>%
    pivot_longer(
        cols = starts_with("rs"),
        names_to = "SNP",
        values_to = "genotype"
    ) %>%
    # Count occurrences for each combination of variables
    group_by(SNP, genotype, diabetes_status) %>%
    summarise(n = n(), .groups = "drop")
ggplot(meta_data_long, aes(x = genotype, y = n, fill = diabetes_status)) +
    geom_bar(stat = "identity", position = position_dodge(width = 0.7), width = 0.6) +
    geom_text(aes(label = n), vjust = -0.5, position = position_dodge(width = 0.7), size = 3) +
    facet_wrap(~SNP, scales = "free_x", nrow = 2) +
    scale_fill_manual(values = diabetes_palette) +
    theme_classic() +
    theme(
        axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "bottom"
    ) +
    labs(x = "Genotype", y = "Count", fill = "Diabetes Status") +
    ylim(0, 30)
# TODO Make this an upset plot
```

## 1. INS expression is similar in beta-cells of AA vs. TT islet donors (_rs689_)

```{r}
#| eval: false
# seurat_object$tosti_cell_type.protected <- paste(seurat_object$tosti_cell_type, seurat_object$protected, sep = "_")
# Idents(seurat_object) <- "tosti_cell_type.protected"
# beta.de <- FindMarkers(seurat_object, ident.1 = "Beta_TRUE", ident.2 = "Beta_FALSE", verbose = FALSE)
# head(beta.de, n = 40)

rs689_pseudo_seurat_object <- AggregateExpression(seurat_object, assays = "RNA", return.seurat = TRUE, group.by = c("rs689_consensus", "orig.ident", "tosti_cell_type"))
rs689_pseudo_seurat_object$tosti_cell_type.rs689 <- paste(rs689_pseudo_seurat_object$tosti_cell_type, rs689_pseudo_seurat_object$rs689_consensus, sep = "_")
Idents(rs689_pseudo_seurat_object) <- "tosti_cell_type.rs689"
bulk.beta.de <- FindMarkers(rs689_pseudo_seurat_object, ident.1 = "Beta_AA", ident.2 = "Beta_TT", verbose = FALSE)

bulk.beta.de$gene <- rownames(bulk.beta.de)

ggplot(bulk.beta.de, aes(x = avg_log2FC, y = -log10(p_val))) +
    geom_point(size = 0.5, alpha = 0.5) + # Adjust point size and opacity for better visualization
    theme_bw(base_size = 14) + # Use theme_bw with a base font size for better readability
    labs(
        x = "Average Log2 Fold Change",
        y = "-log10(Unadjusted p-value)",
        title = "Beta-cell Differential Expression",
        subtitle = "Comparison of AA v TT"
    ) +
    geom_text_repel(
        aes(label = ifelse(p_val_adj < 0.1, gene, "")), # Label significant genes based on adjusted p-value
        size = 3.5, # Slightly larger text size for readability
        box.padding = 0.35, # Adjust padding around text to avoid overlap
        point.padding = 0.5, # Adjust distance from points to text
        min.segment.length = 0.1, # Minimize the segment length for a cleaner look
        color = "blue", # Change text color to blue for better visibility against the background
        nudge_x = 0.05, # Nudge text a bit to the right for clearer separation from points
        nudge_y = 0.05, # Nudge text up a bit for vertical separation
        direction = "y" # Optimize text placement primarily in the vertical direction
    ) +
    theme(
        plot.title = element_text(face = "bold"), # Bold title for emphasis
        plot.subtitle = element_text(face = "italic"), # Italic subtitle for stylistic differentiation
        axis.title.x = element_text(vjust = -0.2), # Adjust x-axis title position for closer proximity to axis
        axis.title.y = element_text(vjust = 1.5), # Adjust y-axis title position for optimal spacing
        legend.position = "none" # Hide legend (not applicable here but good practice for cleaner plots)
    )

# *Confirmed!* This disproves the prevailing hypothesis that VNTR polymorphism in the promoter region of the INS gene explains genetic allele specific expression and protection from T1D (Pugliese; Polychronakos Nat Genet 1997)

```


Wang et al note that INS expression is associated with A allele of rs689 and C allele of rs3842753. Need to test for this association in our data.

::: {.callout-important}
## Reference genome mismatch

Wang et al report alignment with hg38:

> Single-cell RNAseq reads were aligned to human reference genome **GRCh38.98** using STAR version 2.7.1a (25) with gene read counts obtained using –quantMode GeneCounts. Read count files were aggregated into study-specific read count matrices ***using a custom code***.

However: 

```{r}
# ensembl <- useMart("ENSEMBL_MART_SNP", dataset = "hsapiens_snp", host = "https://www.ensembl.org")
snp_ids <- c("rs689", "rs3842753", "rs3842752")


rsnps::ncbi_snp_query(snp_ids) |>
    dplyr::select(-maf_population) |>
    gt::gt()
```

Data suggests they in fact used GRCh37 in which A is the ref for rs3842753 and the minor allele is C; while in GRCh38, the reference allele is T and the minor is G.

- [NCBI ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/304051/?oq=rs3842753&m=NM_000207.3(INS):c.*22A%3EC)
  
  | HGVS                                            | dbSNP       | gnomAD frequency |
  |-------------------------------------------------|-------------|------------------|
  | NM_000207.3(INS):c.-17-6T>A                     | rs689       | 0.59334          |
  | NM_000207.3(INS):c.*22A>C                       | rs3842753   | 0.59307          |
  | NM_000207.3(INS):c.*9C>T                        | rs3842752   | 0.16663          |


- [NCBI SNP](https://www.ncbi.nlm.nih.gov/snp/rs3842753)
  
  | Source        | Allele | Frequency   | Count       | Total     |
  |---------------|--------|-------------|-------------|-----------|
  | TOPMED        | G      | 0.587468    | 155497      | 264690    |
  | GnomAD_exome  | G      | 0.733402    | 176653      | 240868    |
  | GnomAD        | G      | 0.593068    | 83001       | 139952    |

:::

b) Wang et al: rs3842753 C allele (**minor allele**) associates with increased INS expression in single b cells and b-like cells

We test for the **G** allele:

```{r}
#| eval: false
# seurat_object$tosti_cell_type.rs3842753 <- paste(seurat_object$tosti_cell_type, seurat_object$rs3842753_consensus, sep = "_")
# Idents(seurat_object) <- "tosti_cell_type.rs3842753"
# beta.de <- FindMarkers(seurat_object, ident.1 = "Beta_TT", ident.2 = "Beta_TG", verbose = FALSE)
# head(beta.de, n = 40)

rs3842753_pseudo_seurat_object <- AggregateExpression(seurat_object, slot = "counts", assays = "RNA", return.seurat = TRUE, group.by = c("rs3842753_consensus", "orig.ident", "tosti_cell_type"))
rs3842753_pseudo_seurat_object$tosti_cell_type.rs3842753 <- paste(rs3842753_pseudo_seurat_object$tosti_cell_type, rs3842753_pseudo_seurat_object$rs3842753_consensus, sep = "_")
Idents(rs3842753_pseudo_seurat_object) <- "tosti_cell_type.rs3842753"
bulk.beta.de <- FindMarkers(rs3842753_pseudo_seurat_object, ident.1 = "Beta_TT", ident.2 = "Beta_TG", verbose = FALSE)

bulk.beta.de$gene <- rownames(bulk.beta.de)

ggplot(bulk.beta.de, aes(x = avg_log2FC, y = -log10(p_val))) +
    geom_point(size = 0.5, alpha = 0.5) + # Adjust point size and opacity for better visualization
    theme_bw(base_size = 14) + # Use theme_bw with a base font size for better readability
    labs(
        x = "Average Log2 Fold Change",
        y = "-log10(Unadjusted p-value)",
        title = "Beta-cell Differential Expression",
        subtitle = "Comparison of TT (→) v TG (←)"
    ) +
    geom_text_repel(
        aes(label = ifelse(p_val_adj < 0.1, gene, "")), # Label significant genes based on adjusted p-value
        size = 3.5, # Slightly larger text size for readability
        box.padding = 0.35, # Adjust padding around text to avoid overlap
        point.padding = 0.5, # Adjust distance from points to text
        min.segment.length = 0.1, # Minimize the segment length for a cleaner look
        color = "blue", # Change text color to blue for better visibility against the background
        nudge_x = 0.05, # Nudge text a bit to the right for clearer separation from points
        nudge_y = 0.05, # Nudge text up a bit for vertical separation
        direction = "y" # Optimize text placement primarily in the vertical direction
    ) +
    theme(
        plot.title = element_text(face = "bold"), # Bold title for emphasis
        plot.subtitle = element_text(face = "italic"), # Italic subtitle for stylistic differentiation
        axis.title.x = element_text(vjust = -0.2), # Adjust x-axis title position for closer proximity to axis
        axis.title.y = element_text(vjust = 1.5), # Adjust y-axis title position for optimal spacing
        legend.position = "none" # Hide legend (not applicable here but good practice for cleaner plots)
    )

```

```{r}
# Fetch data
data <- FetchData(seurat_object, vars = c("tosti_cell_type", "INS", "rs3842753_consensus")) |> dplyr::filter(tosti_cell_type == "Beta")

ggplot(data, aes(x = as.factor(rs3842753_consensus), y = INS, fill = as.factor(rs3842753_consensus))) +
    geom_jitter(width = 0.2, alpha = 0.1, size = 0.1) +
    geom_boxplot() +
    labs(x = "rs3842753 allele", y = "Expression of INS", title = "INS Expression in Beta Cells - rs3842753") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(
        values = c("TT" = "red", "TG" = "grey"),
        labels = c("TRUE" = "T/T", "FALSE" = "T/G")
    ) +
    theme(legend.position = "none") +
    stat_compare_means(aes(group = rs3842753_consensus), method = "wilcox.test", label = "p.signif", label.x.npc = "center", label.y.npc = "top") +
    stat_summary(aes(label = sprintf("%.2f", ..y..)), fun = mean, geom = "text", vjust = -1.5, hjust = 1.5, size = 4, colour = "blue")
```

```{r}
# Fetch data
data <- FetchData(seurat_object, vars = c("tosti_cell_type", "INS", "rs689_consensus")) |> dplyr::filter(tosti_cell_type == "Beta")

# Assuming 'data' is your dataframe with 'INS', 'protected' columns, and 'protected' is a factor
ggplot(data, aes(x = as.factor(rs689_consensus), y = INS, fill = as.factor(rs689_consensus))) +
    geom_jitter(width = 0.2, alpha = 0.1, size = 0.1) +
    geom_boxplot() +
    labs(x = "rs689 allele", y = "Expression of INS", title = "INS Expression in Beta Cells - rs689") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(
        values = c("AA" = "red", "TT" = "grey"),
        labels = c("TRUE" = "A/A", "FALSE" = "T/T")
    ) +
    theme(legend.position = "none") +
    stat_compare_means(aes(group = rs689_consensus), method = "wilcox.test", label = "p.signif", label.x.npc = "center", label.y.npc = "top") +
    stat_summary(aes(label = sprintf("%.2f", ..y..)), fun = mean, geom = "text", vjust = -1.5, hjust = 1.5, size = 4, colour = "blue")
```

```{r}
# Fetch data
data <- FetchData(seurat_object, vars = c("tosti_cell_type", "INS", "rs3842752_consensus")) |> dplyr::filter(tosti_cell_type == "Beta")

# Assuming 'data' is your dataframe with 'INS', 'protected' columns, and 'protected' is a factor
ggplot(data, aes(x = as.factor(rs3842752_consensus), y = INS, fill = as.factor(rs3842752_consensus))) +
    geom_jitter(width = 0.2, alpha = 0.1, size = 0.1) +
    geom_boxplot() +
    labs(x = "rs3842752 allele", y = "Expression of INS", title = "INS Expression in Beta Cells - rs3842752") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(
        values = c("GA" = "red", "GG" = "grey"),
        labels = c("TRUE" = "G/A", "FALSE" = "G/G")
    ) +
    theme(legend.position = "none") +
    stat_compare_means(aes(group = rs3842752_consensus), method = "wilcox.test", label = "p.signif", label.x.npc = "center", label.y.npc = "top") +
    stat_summary(aes(label = sprintf("%.2f", ..y..)), fun = mean, geom = "text", vjust = -1.5, hjust = 1.5, size = 4, colour = "blue")
```

## Spliced INS mRNA is decreased in S/P vs. S/S beta-cells 

::: {.callout-warning appearance="simple"}

## Kallisto analysis still going...

Counts completed; analysis underway

:::

Yes: Great; No: Check in islets with XBPs (XBPs is the prototypic product of XBP digested by IRE1a activated because of ER stress; i.o.w., XBPs marks an ER-stressed (beta-)cell)

Yes: Great; No: check within single beta-cells of S/P with stress (XBPs) vs. unstressed (only XBP) whether P chromosome product has less than S chromosome product; 

Yes: Great; compare with spliced INS mRNA of stressed beta cells of S/P donors with those of S/S donors. Yes: Super; No? no worries. 

## Markers of ER stress are up-regulated in islets of S/S donors compared to S/P donors

a. Check for candidate genes: we expect genes like IRE1 (ERN1), ATF6, XBP1, PERK (EIF2AK3), BIP (HSPA5), CHOP (DDIT3), PSMB10 and INS to be lower expressed in beta cells with the protective SNP. 

::: {.callout-note}

Here we are plotting expression of the ER stress genes-of-interest in beta cells of S/P vs. S/S across donors (peeudobulk). In most cases, higher expression of stress markers is observed in the "Protected" group, however none is significant.

:::

```{r}
#| eval: false
protected_pseudo_seurat_object <- AggregateExpression(seurat_object, assays = "RNA", return.seurat = T, group.by = c("protected", "orig.ident", "tosti_cell_type"))

data <- FetchData(protected_pseudo_seurat_object, vars = c("tosti_cell_type", "protected", er_genes_of_interest)) |>
    dplyr::filter(tosti_cell_type == "Beta")
data_long <- pivot_longer(data, cols = all_of(er_genes_of_interest), names_to = "Gene", values_to = "Expression")
# Plot
ggplot(data_long, aes(x = as.factor(protected), y = Expression, fill = as.factor(protected))) +
    geom_jitter(width = 0.2, alpha = 0.35, size = 0.5) +
    geom_boxplot(alpha = 0.5) +
    labs(x = "Protected", y = "Expression", title = "ER Stress Gene Expression in Beta Cells") +
    facet_wrap(~Gene, scales = "free_y") + # Separate plot for each gene
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(
        values = c("TRUE" = "red", "FALSE" = "grey"),
        labels = c("TRUE" = "S/P", "FALSE" = "S/S")
    ) +
    theme(legend.position = "none") +
    stat_compare_means(aes(group = protected), method = "wilcox.test", p.adjust.method = "BH", label = "p.signif", label.x.npc = "center", label.y.npc = "center") +
    stat_summary(fun = mean, geom = "point", shape = 16, size = 3, color = "black", fill = "blue") + # Add mean dots
    stat_summary(aes(label = sprintf("%.2f", ..y..)), fun = mean, geom = "text", vjust = -1.5, hjust = 1.5, size = 4, colour = "blue") # Add labels for means

```

b. Check for altered genes (volcano plot) of S/S vs. S/P beta-cells

::: {.callout-note}

Here we formally test for differential expression, pseudo-bulked by cell type and donor.

:::

```{r }
#| eval: false
protected_pseudo_seurat_object$tosti_cell_type.protected <- paste(protected_pseudo_seurat_object$tosti_cell_type, protected_pseudo_seurat_object$protected, sep = "_")
Idents(protected_pseudo_seurat_object) <- "tosti_cell_type.protected"
bulk.beta.de <- FindMarkers(protected_pseudo_seurat_object, ident.1 = "Beta_TRUE", ident.2 = "Beta_FALSE", verbose = FALSE)

bulk.beta.de$gene <- rownames(bulk.beta.de)

ggplot(bulk.beta.de, aes(x = avg_log2FC, y = -log10(p_val))) +
    geom_point(size = 0.5, alpha = 0.5) + # Adjust point size and opacity for better visualization
    theme_bw(base_size = 14) + # Use theme_bw with a base font size for better readability
    labs(
        x = "Average Log2 Fold Change",
        y = "-log10(Unadjusted p-value)",
        title = "Beta-cell Differential Expression",
        subtitle = "Comparison of Protected (->) v susceptible (<-)"
    ) +
    geom_text_repel(
        aes(label = ifelse(p_val_adj < 0.1, gene, "")), # Label significant genes based on adjusted p-value
        size = 3.5, # Slightly larger text size for readability
        box.padding = 0.35, # Adjust padding around text to avoid overlap
        point.padding = 0.5, # Adjust distance from points to text
        min.segment.length = 0.1, # Minimize the segment length for a cleaner look
        color = "blue", # Change text color to blue for better visibility against the background
        nudge_x = 0.05, # Nudge text a bit to the right for clearer separation from points
        nudge_y = 0.05, # Nudge text up a bit for vertical separation
        direction = "y" # Optimize text placement primarily in the vertical direction
    ) +
    theme(
        plot.title = element_text(face = "bold"), # Bold title for emphasis
        plot.subtitle = element_text(face = "italic"), # Italic subtitle for stylistic differentiation
        axis.title.x = element_text(vjust = -0.2), # Adjust x-axis title position for closer proximity to axis
        axis.title.y = element_text(vjust = 1.5), # Adjust y-axis title position for optimal spacing
        legend.position = "none" # Hide legend (not applicable here but good practice for cleaner plots)
    )
```

::: {.callout-note}

No genes were significantly differentially expressed between S/S and S/P beta-cells in the current analysis.
:::

```{r}
#| eval: false
library(gt)
de_markers_sorted <- bulk.beta.de %>%
    arrange(p_val, desc(abs(avg_log2FC))) |>
    select(gene, everything()) |>
    head(n = 20)

de_markers_sorted %>%
    gt() %>%
    tab_header(
        title = "Top 20 Differential Expression Markers",
        subtitle = "Sorted by p-value and absolute logFC"
    ) %>%
    cols_label(
        p_val = "P-value",
        avg_log2FC = "Average Log2 Fold Change",
        pct.1 = "Percent detected Protected",
        pct.2 = "Percent detected susceptible",
        p_val_adj = "Adjusted P-value",
        gene = "Gene"
    ) %>%
    fmt_number(
        columns = c("p_val", "avg_log2FC", "p_val_adj"),
        decimals = 4
    ) %>%
    fmt_percent(
        columns = c("pct.1", "pct.2"),
        decimals = 2
    ) %>%
    tab_options(
        table.width = pct(100),
        data_row.padding = px(5)
    ) %>%
    cols_align(
        align = "center",
        columns = everything()
    )

```

c. Check for affected pathways: Cellular responses to stress, Unfolded Protein Response (UPR), Class I MHC mediated antigen processing & presentation, Insulin processing, Cellular Senescence, Regulation of insulin secretion (from reactome.org)?

::: {.callout-note}

No pathway with significant enrichment was found in the current analysis.
:::

```{r }
#| eval: false
require(clusterProfiler)
require(enrichplot)
cluster_profiler_out <- clusterProfiler::GSEA(de_markers, TERM2GENE = cell_types, scoreType = "pos")
enriched_plot <- if (nrow(enriched_cell_types@result) == 0) grid::textGrob("No enriched cell types found") else barplot(enriched_cell_types, showCategory = 10)

```

```{r }
#| eval: false
# rejoin the layers in the sketched assay this is required to perform differential expression
joined_seurat_object <- seurat_object
joined_seurat_object[["sketch"]] <- JoinLayers(seurat_object[["sketch"]])
Idents(joined_seurat_object) <- "tosti_cell_type"
beta_markers <- FindMarkers(object = joined_seurat_object, ident.1 = "Beta")
beta_markers |>
    filter(p_val_adj < 0.05) |>
    head(30)
beta_markers_protected <- FindMarkers(joined_seurat_object, ident.1 = "TRUE", group.by = "protected", subset.ident = "Beta")
```

## Bonus: Beta-cells are more stressed than alpha cells

a. Pathway enrichment analysis to see which pathways are upregulated or downregulated in S/S vs. S/P per islet cell-type-cluster. 

b. In the UMAP, please overlay differentially expressed ER stress genes to see whether beta cells have more stress-related transcripts than for example alpha cells and acinar cells.

```{r }
#| fig-cap: "Tosti et al cell types"
#| column: margin
DimPlot(seurat_object, group.by = "tosti_cell_type", label = TRUE, cols = cell_type_palette) + NoLegend() + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank())
```

```{r }
#| fig-width: 10
#| fig-height: 10
#| column: page
# FeaturePlot(seurat_object, features = er_genes_of_interest, split.by = "protected", keep.scale = "feature") + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank())
seurat_object_protected <- subset(seurat_object, subset = protected == "TRUE")
seurat_object_susceptible <- subset(seurat_object, subset = protected == "FALSE")
pplot <- plot_density(seurat_object_protected, features = er_genes_of_interest) + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank()) & NoLegend()
splot <- plot_density(seurat_object_susceptible, features = er_genes_of_interest) + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank()) & NoLegend()
```

::: {.panel-tabset}

### Protected

```{r}
pplot
```

### susceptible

```{r}
splot
```

:::


## Extra Bonus: T1D islets are more stressed than T2D islets and non-diabetic islets

Yes: Done! No: Check beta-cells only. No? Take a Guinness. Yes: Take a few more; apply for CSO position.

```{r }
#| eval: false
Beta_cells <- subset(seurat_object, subset = tosti_cell_type == "Beta")
# Fetch data
data <- FetchData(Beta_cells, vars = c("INS", "protected", "diabetes_status"))

```

